rust/arrow/src/json/writer.rs

```diff
@@ -0,0 +1,301 @@
+            batch.columns().iter().enumerate().for_each(|(j, col)| {
+                let col_name = schema.field(j).name();
+                set_column_for_json_rows(&mut rows[base..], row_count, col, col_name);
+            });
+            base += row_count;
+        });
+    }
+
+    rows
+}
+
+/// A JSON writer
+#[derive(Debug)]
+pub struct Writer<W: Write> {
+    writer: BufWriter<W>,
+}
+
+impl<W: Write> Writer<W> {
+    pub fn new(writer: W) -> Self {
+        Self::from_buf_writer(BufWriter::new(writer))
+    }
+
+    pub fn from_buf_writer(writer: BufWriter<W>) -> Self {
+        Self { writer }
+    }
+
+    pub fn write_row(&mut self, row: &Value) -> Result<()> {
+        self.writer.write_all(&serde_json::to_vec(row)?)?;
+        self.writer.write_all(b"\n")?;
+        Ok(())
+    }
+
+    pub fn write_batches(&mut self, batches: &[RecordBatch]) -> Result<()> {
+        for row in record_batches_to_json_rows(batches) {
+            self.write_row(&Value::Object(row))?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs::{read_to_string, File};
+    use std::sync::Arc;
+
+    use crate::json::reader::*;
+
+    use super::*;
+
+    #[test]
+    fn write_simple_rows() {
+        let schema = Schema::new(vec![
+            Field::new("a", DataType::Int32, false),
+            Field::new("b", DataType::Utf8, false),
+        ]);
+
+        let a = Int32Array::from(vec![1, 2, 3, 4, 5]);
+        let b = StringArray::from(vec!["a", "b", "c", "d", "e"]);
+
+        let batch =
+            RecordBatch::try_new(Arc::new(schema), vec![Arc::new(a), Arc::new(b)])
+                .unwrap();
+
+        let mut buf = Vec::new();
+        {
+            let mut writer = Writer::new(&mut buf);
+            writer.write_batches(&vec![batch]).unwrap();
+        }
+
+        assert_eq!(
+            String::from_utf8(buf).unwrap(),
+            r#"{"a":1,"b":"a"}
+{"a":2,"b":"b"}
+{"a":3,"b":"c"}
+{"a":4,"b":"d"}
+{"a":5,"b":"e"}
+"#
+        );
+    }
+
+    fn test_write_for_file(test_file: &str) {
+        let builder = ReaderBuilder::new()
+            .infer_schema(None)
+            .with_batch_size(1024);
+        let mut reader: Reader<File> = builder
+            .build::<File>(File::open(test_file).unwrap())
+            .unwrap();
+        let batch = reader.next().unwrap().unwrap();
+
+        let mut buf = Vec::new();
+        {
+            let mut writer = Writer::new(&mut buf);
+            writer.write_batches(&vec![batch]).unwrap();
+        }
+
+        let result = String::from_utf8(buf).unwrap();
+        let expected = read_to_string(test_file).unwrap();
+        for (r, e) in result.lines().zip(expected.lines()) {
+            assert_eq!(
+                serde_json::from_str::<Value>(r).unwrap(),
+                serde_json::from_str::<Value>(e).unwrap()
+            );
+        }
+    }
+
+    #[test]
+    fn write_basic_rows() {
+        test_write_for_file("test/data/basic.json");
```

Comment:
AFAI understand this will populate a file and not delete it at the end.

Would it be possible to write to a byte stream or something instead of a file? Alternatively, use `tmp` or something?

Reply:
This function reads json test file from existing test data, serializes it to an in memory byte stream, then use that result to compare against the same test file from disk for diff. So there should be no disk write involved.